onnxruntime/core/optimizer/identical_children_consolidation.cc

```diff
@@ -0,0 +1,46 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "core/graph/graph_utils.h"
+#include "identical_children_consolidation.h"
+namespace onnxruntime {
+Status IdenticalChildrenConsolidation::ApplyImpl(Graph& graph, bool& modified, int  /*graph_level*/, const logging::Logger&  /*logger*/) const {
+  GraphViewer const graph_viewer(graph);
+  for (auto node_index : graph_viewer.GetNodesInTopologicalOrder()) {
+    Node* node = graph.GetNode(node_index);
+    if (IsSupportedParentNode(node)) {
+      auto identical_children_indexes = GetIdenticalChildrenIndexes(node);
+      if(identical_children_indexes.empty()) { continue;}
+      auto *first_child = graph.GetNode(identical_children_indexes[0]);
+      for (size_t i = 1; i < identical_children_indexes.size(); i++) {
+        auto *other_child = graph.GetNode(identical_children_indexes[i]);
+        //Currently Only support single output
+        graph_utils::ReplaceDownstreamNodeInput(graph, *other_child, 0, *first_child, 0);
+        graph_utils::RemoveNode(graph,*other_child);
+        modified = true;
+      }
+    }
+  }
+  return Status::OK();
+}
+
+std::vector<NodeIndex> IdenticalChildrenConsolidation::GetIdenticalChildrenIndexes(Node *node) const {
+  unordered_set<NodeIndex> identical_set;
+  for (auto i = node->OutputEdgesBegin(); i != std::prev(node->OutputEdgesEnd(),1); ++i) {
+    if(supported_children_optypes.count(i->GetNode().OpType()) > 0) {
+      for (auto j = std::next(i); j != node->OutputEdgesEnd(); ++j) {
+        if (i->GetNode().OpType() == j->GetNode().OpType()) {
+          identical_set.insert({i->GetNode().Index(), j->GetNode().Index()});
+        }
+      }
```

Comment:
This logic is not enough to determine the child nodes are identical. Q/DQ has more than one input. You also need to check scale and zeropoint

Reply:
There is no need of 2 loops to do this. You can group all identical nodes by:
1. same op type; 2. use concat of all input name as key (for constant input, use a special identifier + constant value to represent the input name); 3. group with a key value map.